text/3778-inline-intents.md

```diff
@@ -0,0 +1,189 @@
+happens to get inlined, then it'll "always" inline that callee too, which might
+not be what was actually desired.
+
+At the same time, sometimes it's useful to put `inline` on things to make the
+definition available to to LLVM, but where it probably shouldn't actually be
+inlined in general, only in particular special cases (perhaps when one of the
+arguments is a small constant, for example).
+
+It would thus be nice to give additional options to the user to let them
+describe *why* they wanted a function inlined, and hopefully be able to make
+better decisions in the backend as a result.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In most cases, plain `#[inline]` is fine, especially with PGO and LTO.
+However, if you've measured and things are making a poor choice, there are some
+options you can use to hint the compiler towards what you want.
+
+## `inline(trampoline)`
+
+This is intended for functions which quickly "bounce" the caller off to some
+other implementation, after doing some initial checks or transformations.
+
+For example, this is useful in a safe function which does some safety checks,
+then calls an `unsafe` version of the function to do the actual work.  Or maybe
+it's a function with a common trivial path, but which sometimes needs to call
+out to a more complicated version, like how `Vec::push` is usually trivial but
+occasionally needs to reallocate.
+
+## `inline(rarely)`
+
+This is intended for functions which normally shouldn't be inlined, but where
+exceptions exist so you don't want full `inline(never)`.
+
+For example, maybe this is a vectorized loop that you wouldn't want copied into
+every caller, but you know that using it for short arrays is common, and thus
+want the back-end to be able to inline and fully unroll it in those cases.
+
+## In combination
+
+These can work particularly well together.
+
+For example, perhaps the public function is `inline(trampoline)` so that the
+initial `NonZero::new` check can be inlined into the caller (where it's more
+easily optimized out) but that calls a private `inline(rarely)` function which
+takes `NonZero` to avoid needing extra checks internally.
+
+Or perhaps you write an `inline(trampoline)` function that picks a strategy
+based on the types and arguments, then dispatches to one of many possible
+`inline(rarely)` implementations.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Like `inline`, these "do nothing" in a spec sense.
+
+The only language change is thus in allowing the two new tokens in the attribute
+(in addition to `never` and `always`).
+
+## Implementation options
+
+⚠ These are not normative, just to illustrate possibilities. ⚠
+
+In LLVM, `#[inline]` sets the [`inlinehint` function attribute](https://llvm.org/docs/LangRef.html#function-attributes),
+so `inline(rarely)` could skip doing that, and thus comparatively slightly
+discourage inlining it.
```

Comment:
Even without `inlinehint` LLVM's inlining heuristic can be mostly characterized as "yes", so I'm not sure if I'd describe this behavior as "inline rarely". This is not just a naming concern -- I don't know off-hand where I'd use an attribute that works this way. Not adding any `#[inline]` attribute and turning on ThinLTO mostly covers the "I don't want to encourage inlining in general but it's fine if it happens" scenarios for me without the costs of emitting multiple copies of the function in different CGUs.

Reply:
Are you actually aiming at rustc's MIR inliner what you're aiming at when proposing this? Not LLVM's inliner? I guess I don't understand what effect this option is supposed to achieve, what gap it's supposed to fill. Would it make the callee's body *available* for inlining in principle even without LTO (by emitting it into multiple CGUs), similar to `#[inline]` but without the extra encouragement for actually inlining it? Or would `#[inline(rarely)]` only make the inliner less eagier in cases where an unannotated callee would be available for inlining anyway, similar but less extreme to `#[inline(never)]`?